app/models/mixins/authentication_mixin.rb

```diff
@@ -2,10 +2,49 @@ module AuthenticationMixin
   extend ActiveSupport::Concern
 
   included do
+    # There are some dirty, dirty Ruby/Rails reasons why this method needs to
+    # exist like this, and I will try and explain:
+    #
+    # First, we need the extra +.where+ here because if this is used in a
+    # nested SELECT statement (aka: virtual_delegate), the +resource_type+
+    # check is dropped on the floor and not included in the subquery.  As a
+    # result, it will just pickup the first record to match the
+    # relationship_id column and that could be associated with any OTHER object
+    # with an Authentication record.
+    #
+    # For example, you get back an authentication record for an EMS record when
+    # you are looking up on for a Host.
+    #
+    # Secondly, the reason this method exists, and is FIRST (prior to the
+    # +has_one+ that uses it) is it needs to be defined prior to the
+    # relationship method that calls it.  Also, it needs to be a method so the
+    # proper local variable, +authentication_mixin_relation+, can be defined
+    # with the class that the +resource_type+ needs to match and remain in
+    # scope for the Proc below.  If it is a class variable or done in other
+    # fashion, it will be overwritten whenever this module is included
+    # elsewhere.
+    #
+    # Finally, the use of a prepared statement is also for some reason required
+    # over the Hash syntax since otherwise the following is ERROR is produced
+    # when doing a nested SELECT:
+    #
+    #     PG::ProtocolViolation: ERROR:  bind message supplies 0 parameters,
+    #     but prepared statement "" requires 1 (ActiveRecord::StatementInvalid)
+    #
+    def self.authentication_status_severity_level_filter
+      # required to be done here so it is in scope of the Proc below
+      authentication_mixin_relation = name
+
+      proc do
+        where('"authentications"."resource_type" = ?', authentication_mixin_relation)
+          .order(Authentication::STATUS_SEVERITY_AREL.desc)
+      end
```

Comment:
Minor point, but would this be clearer to just have this method return the relation (like a normal scope method), and then down below just do

```ruby
-> { authentication_status_severity_level_filter }
```

Reply:
Oh, no, because the proc ends up getting executed in the context of the `relation`, not the class that is mixing this in.  So having it as a method is required because otherwise `name` isn't in the proper scope (the class including this versus the relation at the time the `proc` ends up getting executed).

Reply:
For what it is worth, I did almost say "yes" to your question, but I already tried it in a different form, by using a `class` variable for `authentication_mixin_relation`, so it looked something like this:

```ruby
@@authentication_mix_relation = self.name

has_one  :authentication_status_severity_level,
         -> {  where('"authentications"."resource_type" = ?', @@authentication_mixin_relation).order(Authentication::STATUS_SEVERITY_AREL.desc) },
         authentication_status_severity_level_filter,
         :as         => :resource,
         :inverse_of => :resource,
         :class_name => "Authentication"
```

But that just ends up being hard coded as the last class the happened to include `AuthenticationMixin` into it, so in my tests, it ended up always being `Vm`.

Using the method ensures that at the time the `has_one` is executed, it uses the class that is including it and Ruby has stored the scope for the proc provided.  I can't think of another way to do this.